saleor/graphql/order/mutations/orders.py

```diff
@@ -560,3 +561,35 @@ def perform_mutation(cls, _root, info, **data):
             cls.get_node_or_error(info, data["id"], only_type=Invoice)
         )
         return super().perform_mutation(_root, info, **data)
+
+
+class UpdateInvoice(ModelMutation):
+    invoice = graphene.Field(Invoice, description="An updated invoice.")
+
+    class Arguments:
+        id = graphene.ID(required=True, description="ID of an invoice to update.")
+        number = graphene.String(description="Invoice number")
+        url = graphene.String(description="URL of an invoice to download.")
+
+    class Meta:
+        description = "Updates an invoice."
+        model = models.Invoice
+        permissions = (OrderPermissions.MANAGE_ORDERS,)
+
+    @classmethod
+    def perform_mutation(cls, _root, info, **data):
+        invoice = cls.get_instance(info, **data)
+        with transaction.atomic():
+            invoice.update_invoice(number=data.get("number"), url=data.get("url"))
+            invoice.status = InvoiceStatus.READY
+            invoice.save()
+            if not (invoice.number and invoice.url):
```

Comment:
I think I already ask but I'm still confused with this - why can't we have `clean_input` where we check what was provided and what's in the model and throw the validation error there? In this line, we've already saved the object in the database. The whole operation can fail and we end up with an invalid state in DB.

Reply:
Sorry, got confused while jumping between branches, yes, that can and should be implemented this way. Changing.